// react-obsidian/dist/src/graph/registry/GraphMiddlewareChain.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var DefaultGraphResolver_1 = __importDefault(require("./DefaultGraphResolver"));
var GraphMiddlewareChain = /** @class */ (function () {
    function GraphMiddlewareChain(defaultMiddleware) {
        if (defaultMiddleware === void 0) { defaultMiddleware = new DefaultGraphResolver_1.default(); }
        this.middlewares = [defaultMiddleware];
    }
    GraphMiddlewareChain.prototype.resolve = function (Graph, props) {
        return this.middlewares[0].resolve(this.middlewares[0].resolveChain, Graph, props);
    };
    GraphMiddlewareChain.prototype.add = function (middleware) {
        this.middlewares.unshift(middleware);
        this.updateResolveChain();
    };
    GraphMiddlewareChain.prototype.updateResolveChain = function () {
        if (this.middlewares.length > 1) {
            this.middlewares[0].setNext(this.middlewares[1]);
        }
    };
    GraphMiddlewareChain.prototype.clear = function () {
        this.middlewares = [new DefaultGraphResolver_1.default()];
    };
    return GraphMiddlewareChain;
}());
exports.default = GraphMiddlewareChain;
