// react-obsidian/dist/src/injectors/class/InjectionMetadata.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var ConstructorArgs_1 = require("./ConstructorArgs");
var InjectionMetadata = /** @class */ (function () {
    function InjectionMetadata() {
        this.injectionMetadataKey = 'injectionMetadata';
        this.injectedConstructorArgsKey = 'injectedConstructorArgsKey';
        this.lateInjectionMetadataKey = 'lateInjectionMetadataKey';
    }
    InjectionMetadata.prototype.getConstructorArgsToInject = function (target) {
        var _a;
        return (_a = Reflect.getMetadata(this.injectedConstructorArgsKey, target)) !== null && _a !== void 0 ? _a : new ConstructorArgs_1.ConstructorArgs();
    };
    InjectionMetadata.prototype.getPropertiesToInject = function (target) {
        return this.getProperties(this.injectionMetadataKey, target);
    };
    InjectionMetadata.prototype.getLatePropertiesToInject = function (target) {
        return this.getProperties(this.lateInjectionMetadataKey, target);
    };
    InjectionMetadata.prototype.saveConstructorParamMetadata = function (target, paramName, index) {
        var argsToInject = this.getConstructorArgsToInject(target);
        argsToInject.add(paramName, index);
        Reflect.defineMetadata(this.injectedConstructorArgsKey, argsToInject, target);
    };
    InjectionMetadata.prototype.savePropertyMetadata = function (target, property) {
        this.saveProperties(this.injectionMetadataKey, this.getPropertiesToInject(target).add(property), target);
    };
    InjectionMetadata.prototype.saveLatePropertyMetadata = function (target, property) {
        this.saveProperties(this.lateInjectionMetadataKey, this.getLatePropertiesToInject(target).add(property), target);
    };
    InjectionMetadata.prototype.saveProperties = function (key, properties, target) {
        Reflect.defineMetadata(key, properties, target);
    };
    InjectionMetadata.prototype.getProperties = function (key, target) {
        var _a;
        return (_a = Reflect.getMetadata(key, target)) !== null && _a !== void 0 ? _a : new Set();
    };
    return InjectionMetadata;
}());
exports.default = InjectionMetadata;
