// react-obsidian/dist/src/Obsidian.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var GraphRegistry_1 = __importDefault(require("./graph/registry/GraphRegistry"));
var LateInjector_1 = __importDefault(require("./injectors/class/LateInjector"));
var ServiceLocatorFactory_1 = __importDefault(require("./graph/ServiceLocatorFactory"));
var Obsidian = /** @class */ (function () {
    function Obsidian() {
    }
    Obsidian.prototype.obtain = function (Graph, props) {
        return ServiceLocatorFactory_1.default.fromGraph(Graph, props);
    };
    Obsidian.prototype.inject = function (target, graph) {
        return LateInjector_1.default.inject(target, graph);
    };
    Obsidian.prototype.addGraphMiddleware = function (middleware) {
        GraphRegistry_1.default.addGraphMiddleware(middleware);
    };
    Obsidian.prototype.clearGraphMiddlewares = function () {
        GraphRegistry_1.default.clearGraphMiddlewares();
    };
    Obsidian.prototype.clearGraphs = function () {
        GraphRegistry_1.default.clearGraphMiddlewares();
        GraphRegistry_1.default.clearAll();
    };
    return Obsidian;
}());
exports.default = Obsidian;
