// @nozbe/watermelondb/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

exports.__esModule = true;
exports.Q = exports.localStorageKey = exports.tableSchema = exports.appSchema = exports.columnName = exports.tableName = exports.associations = exports.Query = exports.Model = exports.Relation = exports.Database = exports.Collection = void 0;

var Q = _interopRequireWildcard(require("./QueryDescription"));

exports.Q = Q;

var _Collection = _interopRequireDefault(require("./Collection"));

exports.Collection = _Collection.default;

var _Database = _interopRequireDefault(require("./Database"));

exports.Database = _Database.default;

var _Relation = _interopRequireDefault(require("./Relation"));

exports.Relation = _Relation.default;

var _Model = _interopRequireWildcard(require("./Model"));

exports.Model = _Model.default;
exports.associations = _Model.associations;

var _Query = _interopRequireDefault(require("./Query"));

exports.Query = _Query.default;

var _Schema = require("./Schema");

exports.tableName = _Schema.tableName;
exports.columnName = _Schema.columnName;
exports.appSchema = _Schema.appSchema;
exports.tableSchema = _Schema.tableSchema;

var _LocalStorage = require("./Database/LocalStorage");

exports.localStorageKey = _LocalStorage.localStorageKey;
