// react-obsidian/dist/transformers/babel-plugin-obsidian/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var unmagler_1 = __importDefault(require("./unmagler"));
var providerArgumentsTransformer = {
    visitor: {
        Program: function (path) {
            path.traverse(internalVisitor);
        },
    },
};
var internalVisitor = {
    ClassMethod: {
        enter: function (_a) {
            var node = _a.node;
            unmagler_1.default.saveClassMethod('Provides', node);
        },
    },
    ClassProperty: {
        enter: function (_a) {
            var node = _a.node;
            unmagler_1.default.saveClassProperty('Inject', node);
            unmagler_1.default.saveClassProperty('LateInject', node);
        },
    },
    Identifier: {
        enter: function (_a) {
            var node = _a.node;
            unmagler_1.default.saveIdentifier('Inject', node);
        },
    },
    TSParameterProperty: {
        enter: function (_a) {
            var node = _a.node;
            unmagler_1.default.saveTSParameterProperty('Inject', node);
        },
    },
};
function plugin() {
    return providerArgumentsTransformer;
}
exports.default = plugin;
